### PR TITLE
fix: Fix pathRewrite in k8s Proxy for WebSocket requests

### PR DESCRIPTION
**Pull request by luchillo17**
## Hey, I just made a Pull Request!

WebSocket requests path were not being rewritten by the proxy properly, now they do.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by luchillo17** [comment]
Having issues doing a changeset, getting this error with yarn `Type Error: Cannot read properties of undefined (reading 'toUpperCase')`

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (mclarke47) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:662]: Are these permissionApi mocks no longer necessary?

Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:662]: Moved up into the `beforeEach` since all tests are allowing.

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:715]: I think you don't need this as you're not using the done callback!
```suggestion
```

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:701]: instead of this comment, maybe we could rename `wsServer` to something that describes this echo behaviour, like `wsEchoServer`?

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:671]: Not a huge deal, but using hardcoded ports for these can potentially hurt portability. Probably these ports will be available in most CI environments, but I could imagine them being in use on a dev machine with a bunch of services on localhost.

If I'm not mistaken, in order to make this more robust you can do like [what supertest does](https://github.com/ladjs/supertest/blob/c1b8f9da7b58a5ffe6bf6a1b5056435a21d3c365/lib/test.js#L48-L49) and listen on port "0", which -- on every OS I've used -- will result in the kernel randomly picking an available port for you.

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:698]: I'm have a low tolerance for comments so I'd say either nothing:
```suggestion
```
or just polish the English in the comment:
```suggestion
        // immediately send feedback to the incoming connection
```

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:662]: ^ very happy to see this by the way!

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:755]: Little nit, not blocking -- I think you can await it without giving a name? Just to save a line really
```suggestion
      await eventPromiseFactory(webSocket, 'open');
```

**Review by jamieklassen — CHANGES_REQUESTED** [review]
Feel free to contradict/discuss my inline comments!

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:671]: I'll have to look at supertest to see how they're getting the port after `listen`, since I will need it for the requests.

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:698]: I think it's obvious what the send in the next line is doing, will remove.

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:753]: I also think it should work, probably had it like this because I was asserting something about the resolved value before creating the `eventPromiseFactory` factory, will use the suggestion and see if it still passes.

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:700]: Indeed, will change.

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:671]: Could be that this is even necessary for the GitHub workflows to pass -- I think [this error](https://github.com/backstage/backstage/actions/runs/5596220840/jobs/10232998280?pr=18716#step:8:28) has happened multiple times:
> ```
> ● KubernetesProxy › WebSocket › should proxy websocket connections
>
>    FetchError: request to http://localhost:9000/proxy/ws failed, reason: connect ECONNREFUSED ::1:9000
> ```

As for getting the port after calling `listen(0)`, it looks like you can do

```typescript
app.listen(0);
const port = app.address().port;
```

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:130]: While we're here, I think this version is a little tighter:
```suggestion
    permissionApi.authorize.mockResolvedValue(
      [{ result: AuthorizeResult.ALLOW }],
```

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts]: Done

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts]: Done

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:696]: Maybe forgot to delete?
```suggestion
```

File: .changeset/chatty-seahorses-juggle.md
Comment (jamieklassen) [inline .changeset/chatty-seahorses-juggle.md:5]: Not blocking, I just kinda like the imperative mood for release notes:
```suggestion
Fixed a bug where the proxy was not rewriting WebSocket request paths properly.
```

**Review by jamieklassen — CHANGES_REQUESTED** [review]
Starting to look really good! Small things, and I notice that my idea [here](https://github.com/backstage/backstage/pull/18716#discussion_r1268148306) may be wrong, as tests on node 18 are [still failing](https://github.com/backstage/backstage/actions/runs/5604445044/jobs/10252485298?pr=18716#step:8:35) with `ECONNREFUSED`...

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:696]: Indeed, testing if using the same server would work, was not giving me a good result.

File: plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts:734]: based on node-fetch/node-fetch#1624, it's probably worth trying at least this:
```suggestion
      const wsProxyAddress = `ws://127.0.0.1:${proxyPort}${proxyPath}${wsPath}`;
```
if not changing all the `localhost`s to `127.0.0.1`s

**Review by jamieklassen — CHANGES_REQUESTED** [review]
Have an idea to fix those CI checks

**Comment by luchillo17** [comment]
Hi @mclarke47 keep seeing an `aws-sdk` error in a couple E2E tests, is it another botched release from them? or still the same release and AWS has yet to fix it?

**Comment by mclarke47** [comment]
I'm not very knowledgable about the aws-sdk tbh, did you raise the issue in discord?

**Comment by luchillo17** [comment]
I did it in the `vmware-spotify` channel, should I do it in a specific one?

**Comment by waldirmontoya25** [comment]
Hi @clareliguori , sorry pinging you directly, but just in case you have an idea.

**Comment by clareliguori** [comment]
There was an issue that seems to match what you're seeing - a fixed version v3.378.0 was released on Jul 26:
https://github.com/aws/aws-sdk-js-v3/issues/5014

I'm not sure why the tests aren't respecting the version in the lockfile though (v3.370.0) and instead seem to be pulling in the broken version v3.377.0.  In any case, re-running the tests might fix it.

**Comment by benjdlambert** [comment]
The e2e's that are failing here don't use the lockfile that we have in the monorepo, as they're meant to test new installations of `npx @backstage/create-app` which will basically pull in all the latest deps and run typechecking etc with those. Looks like this might just need a retrigger now though if the issues has been fixed :tada: thanks @clareliguori! :pray:

**Review by jamieklassen — APPROVED** [review]

**Comment by jamieklassen** [comment]
All checks are green, looks like we still need an approval from a maintainer -- @Rugvip or @benjdlambert?

**Review by Rugvip — APPROVED** [review]
:shipit: !

**Pull request merged**
